File: app/src/main/java/com/guuguo/learnsave/adapter/MeiziAdapter.java
package com.guuguo.learnsave.adapter;

import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.chad.library.adapter.base.BaseQuickAdapter;
import com.chad.library.adapter.base.BaseViewHolder;
import com.guuguo.learnsave.R;
import com.guuguo.learnsave.bean.entity.GankBean;
import com.guuguo.learnsave.config.CustomImageSizeModelFutureStudio;

import java.text.SimpleDateFormat;

public class MeiziAdapter extends BaseQuickAdapter<GankBean> {
    public MeiziAdapter() {
        super(R.layout.item_category_image, null);
    }

    //    HashMap<GankBean, Integer> map=new HashMap<>();
//    WindowManager wm = (WindowManager) MyApplication.getInstance().getSystemService(Context.WINDOW_SERVICE);

    @Override
    protected void convert(BaseViewHolder holder, GankBean gankBean) {
        holder.setText(R.id.who, gankBean.getWho())
                .setText(R.id.date, new SimpleDateFormat("yyyy年MM月dd日")
                        .format(gankBean.getPublishedAt()));
//        if (map.containsKey(gankBean)) {
//            holder.getView(R.id.image).getLayoutParams().height = map.get(gankBean);
//            holder.getView(R.id.image).requestLayout();
//        }
        Glide.with(mContext).load(new CustomImageSizeModelFutureStudio(gankBean.getUrl())).into((ImageView) holder.getView(R.id.image));
//                .getSize((width, height) -> {
//            if (!map.containsKey(gankBean)) {
//            int margin = ((ViewGroup.MarginLayoutParams) holder.getConvertView().getLayoutParams()).leftMargin;
//            int myHeight = (wm.getDefaultDisplay().getWidth() - margin * 4) / 2 / width * height;
//                map.put(gankBean, myHeight);
//            holder.getView(R.id.image).getLayoutParams().height = myHeight;
//            }
//                });
    }

    private static final String TAG = "MeiziAdapter";
}

File: app/src/main/java/com/guuguo/learnsave/adapter/MeiziAdapter.kt
package com.guuguo.learnsave.adapter

import android.view.ViewGroup
import android.widget.ImageView

import com.bumptech.glide.Glide
import com.chad.library.adapter.base.BaseQuickAdapter
import com.chad.library.adapter.base.BaseViewHolder
import com.guuguo.learnsave.R
import com.guuguo.learnsave.model.entity.GankModel

import java.text.SimpleDateFormat

import android.R.attr.width
import android.animation.ObjectAnimator
import android.app.Activity
import android.content.Intent
import android.graphics.Bitmap
import android.media.Image
import android.support.v4.app.ActivityCompat
import android.support.v4.app.ActivityOptionsCompat
import android.util.Log
import android.view.View
import android.widget.AdapterView
import android.widget.TextView
import com.bumptech.glide.DrawableTypeRequest
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.resource.bitmap.GlideBitmapDrawable
import com.bumptech.glide.load.resource.drawable.GlideDrawable
import com.bumptech.glide.request.animation.GlideAnimation
import com.bumptech.glide.request.target.BaseTarget
import com.bumptech.glide.request.target.SimpleTarget
import com.bumptech.glide.request.target.SizeReadyCallback
import com.bumptech.glide.request.target.Target
import com.guuguo.learnsave.app.activity.GankActivity
import com.guuguo.learnsave.app.widget.RatioImageView
import com.guuguo.learnsave.extension.date
import com.guuguo.learnsave.extension.getDateSimply
import com.guuguo.learnsave.util.MEIZI
import com.guuguo.learnsave.util.OmeiziDrawable
import com.guuguo.learnsave.util.TRANSLATE_GIRL_VIEW
import java.io.Serializable

class MeiziAdapter : BaseQuickAdapter<GankModel> {
    constructor() : super(R.layout.item_meizi, null) {
    }

    constructor(data: List<GankModel>) : super(R.layout.item_meizi, data) {
    }

    override fun convert(holder: BaseViewHolder, gankBean: GankModel) {
        val image = holder.getView<View>(R.id.image) as RatioImageView
        val describe = holder.getView<View>(R.id.date)
        describe.visibility = View.GONE
        holder.setText(R.id.date, gankBean.publishedAt?.getDateSimply())
                .addOnClickListener(R.id.image)
 
        
        with(gankBean) {
            if (width > 0 && height > 0) {
                knownImageSize(describe, image, width, height)
            }
        }
        Glide.with(mContext).load(gankBean.url)
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .crossFade()
                .into(object : SimpleTarget<GlideDrawable>() {
                    override fun onResourceReady(resource: GlideDrawable, glideAnimation: GlideAnimation<in GlideDrawable>) {
                        val height = resource.intrinsicHeight
                        val width = resource.intrinsicWidth
                        gankBean.width = width
                        gankBean.height = height

                        image.setImageDrawable(resource)
                        knownImageSize(describe, image, width, height)
                    }
                })
        
    }

    private fun knownImageSize(describe: View, image: RatioImageView, width: Int, height: Int) {
        image.setOriginalSize(width, height)
        describe.visibility = View.VISIBLE
    }

    override fun getItemId(position: Int): Long {
        return getItem(position).url!!.hashCode().toLong()
    }

}

